### ISSUE TITLE
Feature: Scroll mode should be persistent in a scrolled pane

### ISSUE THREAD

**Issue by imsnif** [zellij-org/zellij#638]
When I:

1. Enter scroll mode and scroll up inside a pane
2. Move the focus to another pane, exit scroll mode and do something else
3. Go back to the scrolled up pane

I should be placed in scroll mode again.

This is so that I don't try to scroll and then lose my position because the pane resets back to its bottom most position.

**Comment by imsnif** [comment]
Hey @v0idx - could you link to the parts of the code you mean? It would be easier to discuss them that way. :)

**Comment by v0idx** [comment]
Hey @imsnif of course I can! The mode_info attribute can be found here: https://github.com/zellij-org/zellij/blob/e54dfab40f73bbd0c7e27b0caf525c7cc5af607b/zellij-server/src/screen.rs#L138-L154
My thought was that as there is already an attribute that stores the current mode for a screen, by moving this value to the pane object, you would be able to keep track of the current mode, incl. scrolling, on a pane by pane basis

**Comment by imsnif** [comment]
Hey @v0idx - thanks for the link. So to make sure I understand what you're suggesting: you'd like to also store a specific `pane_mode` in each pane (be it TerminalPane or PluginPane) and then when we focus that pane, we'll do a quick query to see if the pane has a specific mode and then "override" the current mode with it?

**Comment by v0idx** [comment]
Yeah @imsnif that was the general jist of what I was suggesting, does this sound applicable to your issue?

**Comment by a-kenji** [comment]
From looking over it sporadically, this sounds like a good solution to me.

@imsnif, @v0idx

**Comment by imsnif** [comment]
Yeah, that sounds good. My only suggestion is to have it as something like `mode_override` in the Pane object and keeping the original in the Screen. I feel this is the shorter path.

**Comment by v0idx** [comment]
That sounds good, I'll get started on an implementation tonight on my fork of zellij

**Comment by arefem-project** [comment]
Hey @v0idx, I looked into this one as well and never finished it. It was actually a little trickier then it seemed at first. Hopefully I can save you some time here.

You should know that the zellij-client also tracks the mode_info locally before sending changes to the server. So when you go to change the mode_info based on the selected pane inside of the screen (or pane) on the server it de-syncs with the clients local mode_info variable and forces you back into the normal mode. I think in order to fix this, you're going to have to add a ServerToClientMsg::SetModeInfo enum variant to propagate the mode_info back to the client when overriding the mode_info for a specific pane. From what I remember you'll also need to wrap the client side mode_info variable in a mutex to propagate it across client threads, or refactor the client code pretty heavily.

That's where I gave up, because it all seemed like a bit of a hack. I feel like I'm 6 hours too late with this information. I hope you've probably already found a much more elegant solution than me.

**Comment by imsnif** [comment]
@arefem-project - thanks for the detailed writeup. I actually forgot there's an `InputMode` on the client side as well. The issue is a little less straight forward than I initially thought.

That being said, unless I'm missing something I think you can probably get away without syncing the InputMode across client threads since we just have the one right now. So if you create the `ServerToClientMsg::SetModeInfo` variant and handle it, this should work.

**Comment by arefem-project** [comment]
Hey @imsnif, I think we're just misunderstanding each other here. I looked again, and I just meant both the "stdin_handler" (InputHandler) and router threads inside the [main client file](https://github.com/zellij-org/zellij/blob/main/zellij-client/src/lib.rs) feeds into a main event loop, and that loop needs to be connected back to the "stdin_handler" thread via a channel or mutex in order to actually change the "mode" variable. I think my wording and fuzzy memory made it seem like I was talking about syncing state across multiple clients or something much more complicated.

If you feel like the `ServerToClientMsg` variant addition is a good solution, and @v0idx doesn't mind, I'll take another look at this sometime this weekend. If you're making progress on it @v0idx, than feel free to finish it.

**Comment by imsnif** [comment]
Sure, @arefem-project - sounds good. Generally in this codebase we went more for the messaging/channel model rather than the arc/mutex model where possible. Let's discuss this further when there's some code written, I think it'll be easier :)

**Comment by spacemaison** [comment]
Hey @imsnif I implemented the PR we discussed (@arefem-project is also my account). You can take a look at it when you've got the time.

The behavior we wanted was actually a bit trickier to implement from a UX perspective then I initially thought it would be. It feels OK to me now but I'm not sure if others will like it, and I've been playing with it for so long that I may have just gotten used to it. The problem was that by default when you switch into the pane InputMode you can then use the movement keys to switch focus between panes. This makes switching into a pane that's previously been in scroll mode kinda jarring, as it feels like you get stuck suddenly in a different mode. I worked around this by adding a check to ensure that the user is in normal mode before switching modes when entering a scroll pane, this allows for the pane mode to keep moving through the pane that should be in scroll mode.

That introduces a new problem now though, the scroll mode seems a little infectious now. When you switch back to a pane that was in normal mode it stays in scroll mode. This is the behavior of Zellij now, but somehow the change from Normal -> Scroll feels like it implies that Scroll -> Normal should happen.

I've been driving myself crazy with this for half a day now. Maybe you can compile my branch and futz around with it to see if it feels normal to you?

**Comment by imsnif** [comment]
Hey @spacemaison - I replied in the PR

**Comment by valentynkit** [comment]
Hey @imsnif, This still bites on 0.44.1: I scroll a pane up to read something, switch to another pane to act on it, and the first pane snaps back to the bottom. Hard to keep one pane parked in the scrollback while working in another.

I looked before reporting and found #696 had most of this working before it was closed as stale. The code's moved on enough since then that I'd rather do it as a fresh PR than revive that branch, keeping the behavior that was worked out there. I'd like to pick this up and work on it, if that direction sounds right to you.
